source/tokens.rs

```diff
@@ -807,4 +1220,348 @@ mod tests {
         assert_eq!(keywords(b"hello"), Result::Error(Error::Position(ErrorKind::Alt, &b"hello"[..])));
     }
+
+    #[test]
+    fn case_span_new() {
+        let input  = &b"foobar"[..];
+        let output = Span {
+            offset: 0,
+            line  : 1,
+            column: 1,
+            slice : input
+        };
+
+        assert_eq!(Span::new(input), output);
+    }
+
+    #[test]
+    fn case_span_new_at() {
+        let input  = &b"foobar"[..];
+        let output = Span {
+            offset: 1,
+            line  : 2,
+            column: 3,
+            slice : input
+        };
+
+        assert_eq!(Span::new_at(input, 1, 2, 3), output);
+    }
+
+    #[test]
+    fn case_span_empty() {
+        let output = Span {
+            offset: 0,
+            line  : 1,
+            column: 1,
+            slice : &b""[..]
+        };
+
+        assert_eq!(Span::empty(), output);
+    }
+
+    #[test]
+    fn case_span_as_slice() {
+        let input  = Span::new(b"foobar");
+        let output = &b"foobar"[..];
+
+        assert_eq!(input.as_slice(), output);
+    }
+
+    #[test]
+    fn case_span_from() {
+        let input  = &b"foobar"[..];
+        let output = Span::new(input);
+
+        assert_eq!(Span::new(input), output);
+    }
+
+    #[test]
+    fn case_span_length_zero() {
+        let input  = Span::new(b"");
+        let output = 0;
+
+        assert_eq!(input.input_len(), output);
+    }
+
+    #[test]
+    fn case_span_length_many() {
+        let input  = Span::new(b"foobar");
+        let output = 6;
+
+        assert_eq!(input.input_len(), output);
+    }
+
+    #[test]
+    fn case_span_empty_iterator_with_indices() {
+        let input  = Span::new(b"");
+        let output = 0;
+
+        assert_eq!(input.iter_indices().len(), output);
+    }
+
+    #[test]
+    fn case_span_iterator_with_indices() {
+        let input  = Span::new(b"foobar");
+        let output = vec![
+            (0, 'f' as u8),
+            (1, 'o' as u8),
+            (2, 'o' as u8),
+            (3, 'b' as u8),
+            (4, 'a' as u8),
+            (5, 'r' as u8)
+        ];
+
+        let mut accumulator = Vec::new();
+
+        for (index, item) in input.iter_indices() {
+            accumulator.push((index, *item));
+        }
+
+        assert_eq!(accumulator, output);
+    }
+
+    #[test]
+    fn case_span_empty_iterator() {
+        let input  = Span::new(b"");
+        let output = 0;
+
+        assert_eq!(input.iter_elements().len(), output);
+    }
+
+    #[test]
+    fn case_span_iterator_with_elements() {
+        let input  = Span::new(b"foobar");
+        let output = vec![
+            'f' as u8,
+            'o' as u8,
+            'o' as u8,
+            'b' as u8,
+            'a' as u8,
+            'r' as u8
+        ];
+
+        let mut accumulator = Vec::new();
+
+        for item in input.iter_elements() {
+            accumulator.push(*item);
+        }
+
+        assert_eq!(accumulator, output);
```

Comment:
I like your way of writing tests. Declaring `input` and `output` as variables to have the objective of the test in a quick glance 👍 .